dmscripts/generate_framework_master_csv.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+"""Class used to output a framework master csv."""
+import csv
+from sys import stdout
+
+
+class GenerateMasterCSV(object):
+    """"Generate a master csv given a framework."""
+
+    static_fieldnames = ['supplier_id', 'supplier_dm_name', 'application_status', 'declaration_status']
+    lot_name_prefixes = ['draft', 'completed']
+
+    def __init__(self, client, target_framework_slug):
+        """Set up CSV builder with a client, framework and lot details and a placeholder for output.
+
+        :param client: Instance of dmapiclient.data.DataAPIClient
+        :param target_framework_slug: A framework slug ie 'digital-outcomes-and-specialists-2' or 'g-cloud-8'
+        """
+        self.client = client
+        self.target_framework_slug = target_framework_slug
+        self.framework = self.client.get_framework(target_framework_slug)['frameworks']
+        self.lot_names = [i['slug'] for i in self.framework['lots']]
+        self.output = []
+
+    def _get_dynamic_field_names(self):
+        """
+        :return: List of strings. Dynamic field names for CSV.
+        """
+        dynamic_fields = []
+        for lot_name in self.lot_names:
+            for prefix in self.lot_name_prefixes:
+                dynamic_fields.append('_'.join([prefix, lot_name]))
+        return dynamic_fields
+
+    def get_fieldnames(self):
+        """
+        :return: List of strings. All field names for CSV.
+        """
+        return self.static_fieldnames + self._get_dynamic_field_names()
+
+    def populate_output(self):
+        """Method to actually populate our output placeholder."""
+        self._update_with_supplier_data(self.output)
+        self._update_with_lot_data(self.output)
```

Comment:
Having these split up like this is a little funny, especially seeing as these two have to be run in a specific order together for it to work. I'm not sure there's really any value over iterating over the suppliers in a single pass and doing the lot-gathering inside the loop.